File: samples/src/main/java/home/saied/samples/ReadObserver.java
package home.saied.samples;

import androidx.annotation.NonNull;
import androidx.compose.runtime.RecomposeScope;
import androidx.compose.runtime.RecomposeScopeImpl;
import androidx.compose.runtime.collection.IdentityArrayIntMap;

import org.apache.commons.lang3.reflect.FieldUtils;

class ReadObserver {
    static Object[] observations(@NonNull RecomposeScope recomposeScope) throws IllegalAccessException {
        IdentityArrayIntMap identityArrayIntMap = (IdentityArrayIntMap) FieldUtils.getField(RecomposeScopeImpl.class, "trackedInstances", true).get(recomposeScope);
        if (identityArrayIntMap == null)
            return new Object[] {};
        return identityArrayIntMap.getKeys();
    }
}

File: samples/src/main/java/home/saied/samples/ReadObserver.kt
package home.saied.samples

import androidx.compose.runtime.*
import androidx.compose.runtime.snapshots.Snapshot

class ScopeRecorder {
    var recomposeScope by mutableStateOf<RecomposeScope?>(null)
}

@Composable
fun StateObserver(
    onUpdate: (Set<Any>) -> Unit,
    content: @Composable ScopeRecorder.() -> Unit
) {
    val scopeRecorder = remember { ScopeRecorder() }
    content(scopeRecorder)
    val contentScope = scopeRecorder.recomposeScope
    DisposableEffect(key1 = content as Any) {
        val subs = Snapshot.registerApplyObserver { set, _ ->
            val reads =
                if (contentScope != null) ReadObserver.observations(contentScope) else emptyArray()
            if (reads.any { it in set })
                onUpdate(reads.toSet())
        }
        onDispose {
            subs.dispose()
        }
    }
}